tests/test_hexary_trie.py

```diff
@@ -109,3 +121,122 @@ def test_trie_using_fixtures(fixture_name, fixture):
         actual_root = trie.root_hash
 
         assert actual_root == expected_root
+
+
+class KeyAccessLogger(dict):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.read_keys = set()
+
+    def __getitem__(self, key):
+        result = super().__getitem__(key)
+        self.read_keys.add(key)
+        return result
+
+    def unread_keys(self):
+        return set(self) - self.read_keys
+
+
+def test_hexary_trie_saves_each_root():
+    changes = ((b'ab', b'b'*32), (b'ac', b'c'*32), (b'ac', None), (b'ad', b'd'*32))
+    expected = ((b'ab', b'b'*32), (b'ad', b'd'*32))
+    db = {}
+    trie = HexaryTrie(db=db)
+    for key, val in changes:
+        if val is None:
+            del trie[key]
+        else:
+            trie[key] = val
+
+    # access all of the values in the trie, triggering reads for all the database keys
+    # that support the final state
+    flagged_usage_db = KeyAccessLogger(db)
+    flag_trie = HexaryTrie(flagged_usage_db, root_hash=trie.root_hash)
+    for key, val in expected:
+        assert flag_trie[key] == val
+
+    # the trie that doesn't prune will certainly have extra keys that aren't needed by
+    # the state with the final root_hash
+    unread = flagged_usage_db.unread_keys()
+    assert len(unread) > 0
+
+
+@pytest.mark.parametrize(
+    'fixture_name,fixture', FIXTURES,
+)
+def test_hexary_trie_saving_final_root(fixture_name, fixture):
+    remaining, deletes = get_expected_results(fixture)
+
+    for kv_permutation in get_fixture_iterator(fixture):
+        db = {}
+        trie = HexaryTrie(db=db)
+        with trie.squash_changes() as memory_trie:
+            for key, value in kv_permutation:
+                if value is None:
+                    del memory_trie[key]
+                else:
+                    memory_trie[key] = value
+
+            for key in deletes:
+                del memory_trie[key]
+
+        # access all of the values in the trie, triggering reads for all the database keys
+        # that support the final state
+        flagged_usage_db = KeyAccessLogger(db)
+        flag_trie = HexaryTrie(flagged_usage_db, root_hash=trie.root_hash)
+        for key, val in remaining.items():
+            assert flag_trie[key] == val
+
+        # assert that no unnecessary database values were created
+        unread = flagged_usage_db.unread_keys()
+        straggler_data = {k: (db[k], decode_node(db[k])) for k in unread}
+        assert len(unread) == 0, straggler_data
+
+        expected_root = fixture['root']
+        actual_root = trie.root_hash
+        assert actual_root == expected_root
+
+
+def test_hexarie_trie_saves_old_keys_after_batch_save():
```

Comment:
I'm confused by this test's name as it doesn't seem to assert anywhere that the previous value is still present in the trie. I guess it's just missing that check at the end?

Reply:
Yeah, on another review, I think we cover everything in the one below. I'll remove this one.